cpp/include/cudf/fixed_point/floating_conversion.hpp

```diff
@@ -0,0 +1,277 @@
+    return integer_rep;
+  }
+
+  /**
+   * @brief Reinterpret the bits of an integer as floating-point value
+   *
+   * @param integer The integer to cast
+   * @return A floating-point value with bits identical to the input
+   */
+  CUDF_HOST_DEVICE inline static FloatingType bit_cast_to_floating(IntegralType integer)
+  {
+    // Convert back to float
+    FloatingType floating;
+    memcpy(&floating, &integer, sizeof(floating));
+    return floating;
+  }
+
+  /**
+   * @brief Extracts the integral significand of a floating-point number
+   *
+   * @param floating The floating value to extract the significand from
+   * @return The integral significand, bit-shifted to a (large) whole number
+   */
+  CUDF_HOST_DEVICE inline static IntegralType get_base2_value(FloatingType floating)
+  {
+    // Convert floating to integer
+    auto const integer_rep = bit_cast_to_integer(floating);
+
+    // Extract the significand, setting the high bit for the understood 1/2
+    return (integer_rep & mantissa_mask) | understood_bit_mask;
+  }
+
+  /**
+   * @brief Extracts the sign bit of a floating-point number
+   *
+   * @param floating The floating value to extract the sign from
+   * @return The sign bit
+   */
+  CUDF_HOST_DEVICE inline static bool get_is_negative(FloatingType floating)
+  {
+    // Convert floating to integer
+    auto const integer_rep = bit_cast_to_integer(floating);
+
+    // Extract the sign bit:
+    return static_cast<bool>(sign_mask & integer_rep);
+  }
+
+  /**
+   * @brief Extracts the exponent of a floating-point number
+   *
+   * @note This returns INT_MIN for +/-0, +/-inf, NaN's, and denormals
+   * For all of these cases, the decimal fixed_point number should be set to zero
+   *
+   * @param floating The floating value to extract the exponent from
+   * @return The stored base-2 exponent, or INT_MIN for special values
+   */
+  CUDF_HOST_DEVICE inline static int get_exp2(FloatingType floating)
+  {
+    // Convert floating to integer
+    auto const integer_rep = bit_cast_to_integer(floating);
+
+    // First extract the exponent bits and handle its special values.
+    // To minimize branching, all of these special cases will return INT_MIN.
+    // For all of these cases, the decimal fixed_point number should be set to zero.
+    auto const exponent_bits = integer_rep & exponent_mask;
+    if (exponent_bits == 0) {
+      // Because of the understood set-bit not stored in the mantissa, it is not possible
+      // to store the value zero directly. Instead both +/-0 and denormals are represented with
+      // the exponent bits set to zero.
+      // Thus it's fastest to just floor (generally unwanted) denormals to zero.
+      return INT_MIN;
+    } else if (exponent_bits == exponent_mask) {
+      //+/-inf and NaN values are stored with all of the exponent bits set.
+      // As none of these are representable by integers, we'll return the same value for all cases.
```

Comment:
Suppose we converted nan/inf to INT_MIN.  On the conversion back to float, we wouldn't know whether the value was inf/nan or whether it was truly INT_MIN.  This information is currently lost anyway.

Reply:
That is my point. When we go back we don't know which it was. Is this acceptable? I don't know the requirements here.

Reply:
To keep track of it we'd have to add some extra bools to fixed_point to know whether we are actually one of these special, unrepresentable floating-point values.  This functionality doesn't exist for the current conversion code so this PR isn't a change regarding it.  If it turns out that functionality is needed it can be added down the road in a separate PR.

Reply:
I'm fine if there is precedent. That was mainly what I was verifying.